File: sdks/dotnet/src/Models/TargetingRule.cs
using System.Text.Json.Serialization;

namespace PhaseFlag.Models;

/// <summary>
/// A targeting rule with conditions, an optional explicit variation,
/// and an optional percentage rollout.
/// </summary>
public sealed class TargetingRule
{
    /// <summary>
    /// Rule priority. Lower numbers are evaluated first.
    /// </summary>
    [JsonPropertyName("priority")]
    public int Priority { get; set; }

    /// <summary>
    /// All conditions must match (AND logic) for the rule to apply.
    /// </summary>
    [JsonPropertyName("conditions")]
    public List<TargetingCondition> Conditions { get; set; } = [];

    /// <summary>
    /// Explicit variation to serve when the rule matches. Mutually exclusive
    /// with <see cref="PercentageRollout"/>.
    /// </summary>
    [JsonPropertyName("variation_id")]
    public string? VariationId { get; set; }

    /// <summary>
    /// Percentage rollout configuration when no explicit variation is set.
    /// </summary>
    [JsonPropertyName("percentage_rollout")]
    public PercentageRollout? PercentageRollout { get; set; }

    /// <summary>
    /// Optional segment identifier for segment-based targeting.
    /// </summary>
    [JsonPropertyName("segment_id")]
    public string? SegmentId { get; set; }
}

File: sdks/php/src/Models/TargetingRule.php
<?php

declare(strict_types=1);

namespace PhaseFlag\Models;

/**
 * A targeting rule with conditions, an optional explicit variation, and an
 * optional percentage rollout.
 *
 * Rules are evaluated in priority order (lower number = higher priority).
 * All conditions within a rule must match (AND logic) for the rule to apply.
 * When a rule matches it can either select an explicit variation or delegate
 * to a percentage rollout.
 */
final class TargetingRule
{
    /**
     * @param int                      $priority          Rule priority (lower = higher priority).
     * @param TargetingCondition[]     $conditions        Conditions that must all match.
     * @param string|null              $variationId       Explicit variation to serve when matched.
     * @param PercentageRollout|null   $percentageRollout Rollout to use when no explicit variation is set.
     * @param string|null              $segmentId         Optional segment reference.
     */
    public function __construct(
        public readonly int $priority = 0,
        public readonly array $conditions = [],
        public readonly ?string $variationId = null,
        public readonly ?PercentageRollout $percentageRollout = null,
        public readonly ?string $segmentId = null,
    ) {}

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'priority'           => $this->priority,
            'conditions'         => array_map(
                static fn (TargetingCondition $c) => $c->toArray(),
                $this->conditions,
            ),
            'variation_id'       => $this->variationId,
            'percentage_rollout' => $this->percentageRollout?->toArray(),
            'segment_id'         => $this->segmentId,
        ];
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): static
    {
        $conditions = array_map(
            static fn (array $c) => TargetingCondition::fromArray($c),
            $data['conditions'] ?? [],
        );

        $rollout = isset($data['percentage_rollout'])
            ? PercentageRollout::fromArray($data['percentage_rollout'])
            : null;

        return new static(
            priority: (int) ($data['priority'] ?? 0),
            conditions: $conditions,
            variationId: $data['variation_id'] ?? null,
            percentageRollout: $rollout,
            segmentId: $data['segment_id'] ?? null,
        );
    }
}
